### PR TITLE
Airspace clearance (glider sector) feature

### LINKED ISSUE #2375: Allow airspace "clearance" in addition to ACK, to handle a) glider sectors (airspace-in-airspace), and b) actual ATC clearances for airspaces
I would like to propose a feature to introduce "clearances" in airspace handling. This is to handle sectors (glider-allowed airspaces inside other non-allowed airspaces, and of course direct ATC clearances given to you by radio).

(I discussed this previously in #1462 / #1468)

This would be handled parallel to the "acknowledge" status of airspaces, but it is distinct:

  * a clearance does not "disable" the airspace, but makes it a "positive" airspace: i.e. you are good as long as you are **inside** that airspace, even if that airspace is contained in another airspace which you would otherwise not be allowed to fly in
  * a clearance could be for the whole day, but it could also be for "once", which would mean it gets revoked automatically once you leave that airspace
  * Inside the airspace with "clearance" status, the warning logic is reversed such that you get a warning if you get close to leaving the airspace, if the point of exit is inside another airspace for which you do not have set a clearance
  * Ideally, also the infobox status (next AS vertical / horizontal) would also be inverted such that it would show the altitude/distance to exiting the current airspace
  * For an airspace that is not contained in another airspace, "clearance" is basically the same as "acknowledge" status. Nevertheless, we still need both distinct properties, for example in the quite common scenario that a restricted area or danger area is partially inside another restricted airspace: In that case, when that restricted / danger area is NOT active, setting it to "clearance" would allow you to fly into the other restircted airspace as long as you stay within the restricted/danger area - which is NOT what you want. So I believe it is unavoidable to add this additional status to the internal airspace stat table.

The workflow would be this: Already before the start, you can activate clearances for sectors that you know are active on that given day. During the flight, if you get information that another sector is activated, you can add that clearance. If you get an airspace warning for a sector that you have an individual or general clearance for, you get an option to set clearance, in addition to the option to set acknowledged.

One very important aspect is to avoid confusion between the curren "acknowledge" state and the new "clearance" state. We therefore should be careful with the wording. This (and other things) is discussed in many more words in #1468.

### PR DESCRIPTION
**Pull request by groundhogxc**
This PR implements an "airspace clearance" status distinct from the existing "acknowledge"
action. (#2375)

When a pilot has ATC clearance or is operating in a glider sector nested within restricted airspace, they can mark that airspace as "cleared".
Unlike acknowledge (which suppresses warnings for airspace you want to avoid), clearance marks airspace you are allowed to be inside, even if there is another airspace at the same location.

This changes the warning logic: An airspace with a clearance, regardless of its class or type, means that any entry warning into another airspace at a location covered by the cleared airspace is suppressed. In the same way, inside warnings are also suppressed.

On exit of a cleared area, the system alerts when approaching the exit boundary if that point leads into a restricted airspace.

Screenshot 2026-05-21 152553

**Implementation:**

**Up to now** warnings were generated by first checking the current position agains all airspaces to find
INSIDE warnings as well as vertical NEAR warnings. Then, the tajectory is predicted using three
different models (glide, filter and task), and the intersection of the path with all airspaces is tested
to see if any of the paths intersect with any airspace within the warning distance / time.

**Now**, the existing part of the algorithm is still basically the same, but in addition, all intersect predictions
for all prediction methods are initially stored instead of thrown away. That means, if a prediction intersects
a given airspace, the entry and exit pointy as well as the length of the segment in between are stored.

Then, after all warnings have been found, there is a new step, where each of those intersect intervals (usually this 
is a small number below 5 or so) for airspaces that have clearances are *subtracted* from the intervals created
from airspaces with no clearance. Segments shorter than 50m are dropped.

After this step, only the remaining warnings (if any) are shown as warnings on the screen, and only if the 
beginning of the segment is close enough to be below the warning threshold.

In addition to this algorithmic extension, there are the following UI extensions:
- Buttons to grant and revoke clearances for all airspaces in the relevant dialogs and monitors
- Display elements / color coding to show the clearance state in the relevant dialogs and the warning monitor
- A new InsideAirspace widget which shows which arespace you are currently in. It only shows if there is 
a suppressed INSIDE warning - so when you are inside both a cleared airspace and inside another non-cleared 
airspace that would have generated a warning were it not for being covered by the cleared airspace. This 
helps situational awareness in complex airspace environments, and it helps the pilot to realize more quickly 
if the wrong airspace was cleared, in particular if that was a large airspace rather than the sector that
was intended to be cleared.
- The airspace boundary of a cleared airspace is shown like a disabled airspace (boundary only), but when it
is intersecting with a warning airspace, the renderer shows the padding of the warning airspace in the shape
of the original airspace with the cleared airspace shape subtracted.
- In particular,it also shows the warning fill of any warning airspace only in the area that is *not* 
covered by cleared airspace(s).

## Summary by CodeRabbit

* **New Features**
  * Airspace clearance management: set/revoke whole-day clearance from airspace dialogs, warning lists, map items, and bottom-area widget
  * Current airspaces widget: compact bottom-area view showing suppressed/cleared entries with quick actions
  * Nearest-warning calculation updated to account for clearance; altitude-aware clearance gating

* **Visual**
  * Cleared airspaces show a "Cleared" tag and green styling; renderer masks cleared areas appropriately

* **Bug Fixes**
  * Improved intersection/interval edge-case handling and clearance clipping

* **Tests**
  * New unit and integration tests for interval math and end-to-end clearance behavior

* **Documentation**
  * Manual updated with "Airspace Clearances" and filter-dialog clearance column guidance

### CODE REVIEW

**Comment by lordfolken** [comment]
I guess we should rename ACK / ACK Day to Dismiss and "Dismiss for" to make this consistent? What do you think?

**Comment by WolfgangDrescher** [comment]
I think this is also the right place to discuss the integration of #2541. The clearance feature really helps with readability in complex airspaces. Thanks for your work, @groundhogxc!

A few observations I had regarding the content of the airspace InfoBoxes.

Here is my flight in case you want to reproduce it: [example IGC file](https://github.com/user-attachments/files/28283918/2026-04-26.Wolfgang.Drescher.1055397.igc.zip)

1. When I fly toward `CTR Sektor Altdorf`, XCSoar's popup prompts me to Ack/Clear `CTR Lahr`. How can we teach XCSoar to determine which of the two airspaces I'm flying toward is the relevant one? My naive approach would be to always suggest the one with the smaller area first, though this could probably cause issues in more complex airspace situations than my case.
2. When I am inside `CTR Sektor Altdorf` and have cleared that sector, "Nearest Airspace Horizontal" shows `CTR Sektor Offenburg`. I think it should show `CTR Lahr` instead.
3. When I am inside `CTR Sektor Altdorf` (cleared or not, it doesn't matter here), "Nearest Airspace Vertical" shows `---`, but it should show the ceiling of `CTR Sektor Altdorf` since `CTR Lahr` is directly above. See #2541 for more details.
4. Shortly before landing and before exiting `CTR Sektor Altdorf`, "Nearest Airspace Vertical" actually shows +6 m or similar. However, XCSoar is actually displaying -6 m AGL, so I believe this behavior is expected.

Bildschirmfoto 2026-05-27 um 20 07 40


> I guess we should rename ACK / ACK Day to Dismiss and "Dismiss for" to make this consistent? What do you think?

I was thinking about "Ignore warnings".

**Comment by groundhogxc** [comment]
@coderabbitai Full review please. A special focus should be on the algorithms being correct, i.e. no warnings erroneously suppressed or not suppressed. Consider any edge case. A secondary concern (for now) is stability.

**Comment by groundhogxc** [comment]
@coderabbitai: In issue number 4 of the full review: Why is the second subtraction operation idempotent for INSIDE warnings, but not for NEAR warnings? Should it not also be? Is not the same interval applied twice?

File: src/Airspace/NearestAirspace.cpp [hunk 1]
```diff
@@ -131,3 +135,156 @@ NearestAirspace::FindVertical(const MoreData &basic,
...
+
+    const bool is_cleared = warnings.IsCleared(airspace);
+    const bool is_warning =
+      !is_cleared &&
+      airspace.IsActive() &&
+      warnings.IsWarningCapable(airspace) &&
+      !warnings.IsAcked(airspace);
+
+    if (!is_cleared && !is_warning)
+      continue;
```
[the remark this answers is not in the document]
Comment (groundhogxc) [reply src/Airspace/NearestAirspace.cpp:263]: Clearance is not meant to be suppressed by isActive(). If ATC formulates a clearance to the pilot in the form of "cleared up to FL xxx within the boundary of yyy" then it does not matter if yyy is active or not.
